// nestjs-typeorm-paginate/dist/create-pagination.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createPaginationObject = void 0;
const pagination_1 = require("./pagination");
function createPaginationObject(items, totalItems, currentPage, limit, route) {
    const totalPages = Math.ceil(totalItems / limit);
    const hasFirstPage = route;
    const hasPreviousPage = route && currentPage > 1;
    const hasNextPage = route && currentPage < totalPages;
    const hasLastPage = route;
    const symbol = route && new RegExp(/\?/).test(route) ? '&' : '?';
    const routes = {
        first: hasFirstPage ? `${route}${symbol}limit=${limit}` : '',
        previous: hasPreviousPage
            ? `${route}${symbol}page=${currentPage - 1}&limit=${limit}`
            : '',
        next: hasNextPage
            ? `${route}${symbol}page=${currentPage + 1}&limit=${limit}`
            : '',
        last: hasLastPage
            ? `${route}${symbol}page=${totalPages}&limit=${limit}`
            : '',
    };
    const meta = {
        totalItems: totalItems,
        itemCount: items.length,
        itemsPerPage: limit,
        totalPages: totalPages,
        currentPage: currentPage,
    };
    return new pagination_1.Pagination(items, meta, route && routes);
}
exports.createPaginationObject = createPaginationObject;
